### ISSUE TITLE
Implement rebasing

### ISSUE THREAD

**Issue by drewmccormack** [drewmccormack/ensembles#38]
This is how it should work:

When a device first joins, it generates a baseline event with all of the objects in its database as inserts. This event will have a global count of 0.

When two baseline events are discovered, they need to be combined. If one is discovered to be a subset of the other, because all events that it includes are also in the other event, then it can be removed.

If neither baseline is a subset of the other, the union must be taken and put into a new baseline, with the original baselines deleted. We simply go through the inserts in each baseline, keeping one for each object, using global count to choose which one to keep, as usual.

Each baseline belongs to a baseline-history. As long as a baseline is just based on a previous baseline, plus any recent changes, it is in the same baseline history. However, if two disjoint baselines are merged with a union, the new baseline belongs to a new baseline-history.

The reason this is significant is that each event store will have to store the baseline-history id that the Core Data store is based on. If the history id of the current baseline differs from that stored in the event store, the Core Data store needs a rebuild from the new baseline. This involves applying the inserts from the baseline, and all newer change sets, to the existing Core Data store. A baseline can only ever include inserts, and a merge can only ever be a superset, so we should not need to figure out what objects in the current Core Data store need to be deleted or anything like that.

The baseline in a single baseline-history needs to be updated regularly, compacting the new events. 
To find the included events, we use this blog post: http://blog.helftone.com/clear-in-the-icloud/

In short, we decide on a global count that will be the cutoff. We determine the supremum of local revisions, and that becomes the set of events included, as well as the revision numbers assigned to the baseline. The global count used for the baseline is the minimum of the global counts on each device’s most recent event.
